pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/ElasticSearchSink.java

```diff
@@ -18,139 +18,254 @@
+        }
 
-        if (!exists) {
-            CreateIndexRequest cireq = new CreateIndexRequest(elasticSearchConfig.getIndexName());
+        String id = key.toString();
+        if (keySchema != null) {
+            id = stringifyKey(keySchema, key);
+        }
 
-            cireq.settings(Settings.builder()
-               .put("index.number_of_shards", elasticSearchConfig.getIndexNumberOfShards())
-               .put("index.number_of_replicas", elasticSearchConfig.getIndexNumberOfReplicas()));
+        String doc = null;
+        if (value != null) {
+            if (valueSchema != null) {
+                doc = stringifyValue(valueSchema, value);
+            } else {
+                doc = value.toString();
+            }
+        }
 
-            CreateIndexResponse ciresp = getClient().indices().create(cireq, RequestOptions.DEFAULT);
-            if (!ciresp.isAcknowledged() || !ciresp.isShardsAcknowledged()) {
-                throw new RuntimeException("Unable to create index.");
+        if (elasticSearchConfig.isKeyIgnore()) {
+            if (doc == null || Strings.isNullOrEmpty(elasticSearchConfig.getPrimaryFields())) {
+                id = Hex.encodeHexString(record.getMessage().get().getMessageId().toByteArray());
+            } else {
+                try {
+                    // extract the PK from the JSON document
+                    JsonNode jsonNode = objectMapper.readTree(doc);
+                    List<String> pkFields = Arrays.asList(elasticSearchConfig.getPrimaryFields().split(","));
+                    id = stringifyKey(jsonNode, pkFields);
+                } catch (JsonProcessingException e) {
+                    log.error("Failed to read JSON", e);
+                }
             }
         }
+
+        SchemaType schemaType = null;
+        if (record.getSchema() != null && record.getSchema().getSchemaInfo() != null) {
+            schemaType = record.getSchema().getSchemaInfo().getType();
+        }
+        log.debug("recordType={} schemaType={} id={} doc={}",
+                record.getClass().getName(),
+                schemaType,
+                id,
+                doc);
+        return Pair.of(id, doc);
     }
 
-    private URL getUrl() throws MalformedURLException {
-        if (url == null) {
-            url = new URL(elasticSearchConfig.getElasticSearchUrl());
+    public String stringifyKey(Schema<?> schema, Object val) {
+        switch (schema.getSchemaInfo().getType()) {
+            case INT8:
+                return Byte.toString((Byte) val);
+            case INT16:
+                return Short.toString((Short) val);
+            case INT32:
+                return Integer.toString((Integer) val);
+            case INT64:
+                return Long.toString((Long) val);
+            case STRING:
+                return (String) val;
+            case JSON:
+                try {
+                    GenericJsonRecord genericJsonRecord = (GenericJsonRecord) val;
```

Comment:
no need to use non public API classes like GenericJsonRecord

I suggest:
`JsonNode node = (JsonNode) ((GenericRecord) val.getNativeObject());`